### PR TITLE
New App: Roomba

### PR DESCRIPTION
**Pull request by noahpodgurski**
# Roomba

Show status of Roomba. Requires a server (raspbi on local network preferrably) running `index.js` to communicate to the Roomba through MQTT protocol. 

roomba2
roomba3
roomba1

This project may be possible without the use of this server - by calling http requests to the Roomba API straight from the pixlet `roomba.star`. However, this may only be possible on older versions of roomba firmware. As I understand it, the newer firmware's use mqtt protocol to transfer data, so for that you need a server in the middle to interface and convert the flow to http. For more information, see https://github.com/koalazak/dorita980

## Config

`serverIP` IP of the server which has `index.js` running.

`serverPORT` (optional) Port of the server. Defaults to 6565.

`roombaIP` (optional) IP of the Roomba. If left blank, the server will automatically search for the IP, in testing, this adds `500ms` to the request time.

## Server Set up
1. NodeJS is installed
2. Copy contents of `/server`
3. Run `npm i`
4. Run `npm install -g dorita980`
5. Run `get-roomba-password-cloud <iRobot Username> <iRobot Password> [Optional API-Key]`. This takes your iRobot username and password. The same you would use to sign into the app or website.
6. Take the resulting `username/blid` and `password` values and paste into `index.js`
7. Run `npm start` to start the server

# Copilot

### Summary
🧹📄🌐

This pull request adds a new applet for tidbyt that displays the status and battery level of a Roomba device. It consists of four files in the apps/roomba directory: `manifest.yaml`, `README.md`, `roomba.star`, and `server/index.js`, which provide the applet metadata, documentation, logic, and rendering code, and the server code that communicates with the Roomba device. It also adds a `package.json` file in the server directory that specifies the server configuration and dependencies.

> _`Roomba` applet code_
> _Manifest, readme, star, server_
> _Autumn cleaning time_

### Walkthrough
*  Register and document the Roomba applet with the tidbyt app store ([link](https://github.com/tidbyt/community/pull/1509/files?diff=unified&w=0#diff-3d9a567fa727d5124fc43cef03e2e46a06ae716463eb1582d687bdf453083801R1-R8), [link](https://github.com/tidbyt/community/pull/1509/files?diff=unified&w=0#diff-0126febc78e375fcd9e9a553b7ac7057afdc0db53dd588b52a449ed073a1c8b9L1-R25))
* Implement the Roomba applet logic and rendering code ([link](https://github.com/tidbyt/community/pull/1509/files?diff=unified&w=0#diff-e30ae99e3ce331353e74da47892999add51030b44eb57b6beab2815ac34b9187R1-R340))
* Create a Node.js server that communicates with the Roomba device ([link](https://github.com/tidbyt/community/pull/1509/files?diff=unified&w=0#diff-1f878780d2be5838df02364fd8f6c47a523879687c75f134899371f5269665b2L1-R62), [link](https://github.com/tidbyt/community/pull/1509/files?diff=unified&w=0#diff-1fe2761f8879ff4174e3e89d806ee0033f4fe2a01de3a0fed57978fba9233f44R1-R17))

### CODE REVIEW

**Comment by noahpodgurski** [comment]
Updated logo
old
old logo

new
new logo

**Comment by matslina** [comment]
Hey there. Tidbyt apps run in the Tidbyt backend, so they don't have access to local networks. That looks like it would be a showstopper for this app, unless there's some way of exposing the data to the public internet?

**Comment by noahpodgurski** [comment]
> Hey there. Tidbyt apps run in the Tidbyt backend, so they don't have access to local networks. That looks like it would be a showstopper for this app, unless there's some way of exposing the data to the public internet?

So the way it works is you host it on your raspberry pi, and port forward it, and then use the pi's public IP as the server IP. All the processing is done on the server to retrieve the roomba info.

File: apps/roomba/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+id: roomba
+name: Roomba
+summary: Shows status of Roomba
+desc: Shows status of Roomba and Braava (i7/i7+, 980, 960, 900, e5, 690, 675, m6, etc) https://github.com/tidbyt/community/blob/869d527834126fd02989dae3bd8ebeceb355600d/apps/roomba/README.md for setup.
```
Comment (matslina) [inline apps/roomba/manifest.yaml:5]:
```suggestion
desc: Shows status of Roomba and Braava (i7/i7+, 980, 960, 900, e5, 690, 675, m6, etc.) Require setting up a local server to host the data. See https://github.com/tidbyt/community/blob/869d527834126fd02989dae3bd8ebeceb355600d/apps/roomba/README.md for more information.
```

**Comment by matslina** [comment]
Ok, sounds good. Wanted to suggest a bit of an edit to the app description to clarify.

Could you also move the files from `apps/roomba/server/` down into `apps/roomba`? The repository structure expects all files to be in the same single directory.

**Comment by matslina** [comment]
Oh, another thing, wanted to mention that the HTTP module now has caching built in. Check out [this post](https://discuss.tidbyt.com/t/http-caching-client/5129) for more information. It should help simplify the code a little bit, and also future proof against potential future changes to the scoping of the cache module.

**Comment by noahpodgurski** [comment]
> Oh, another thing, wanted to mention that the HTTP module now has caching built in. Check out [this post](https://discuss.tidbyt.com/t/http-caching-client/5129) for more information. It should help simplify the code a little bit, and also future proof against potential future changes to the scoping of the cache module.

@matslina updated with ttl_seconds

**Pull request merged**